architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,442 @@
+
+#### Declarative User-defined types are implemented as an RP 
+
+The proposal for user-defined types creates a new resource provider (Dynamic RP) to host the implementation of all user-defined types. The contract between UCP and a user-defined type is the same as for any full resource provider.
+
+The main advantage of implementing user-defined types as a resource provider is that it's consistent with our current architecture. 
+
+The proxy layer in UCP completes the following steps to route traffic:
+
+- Validate the existence of the destination plane and resource group.
+- Look up the resource provider / resource type and its downstream address.
+- Proxy traffic to the resource provider.
+- Interpret the result and update internal state.
+
+##### Proposal
+
+We should implement user-defined types as a resource provider with dynamic behavior. The new resource-provider could be hosted inside UCP directly or in its own microservice.
+
+### API design (if applicable)
+
+API design will be covered as part of the *Resource Type registration APIs* and *Resource Type schemas* design documents.
+
+### CLI Design (if applicable)
+
+CLI design will be covered as part of the *Tooling support for using user-defined types* design document.
+
+### Implementation Details
+
+N/A for this document. This is an overview. 
+
+### Error Handling
+
+N/A for this document. This is an overview. 
+
+## Test plan
+
+N/A for this document. This is an overview. 
+
+## Security
+
+The user-defined types feature has a few touchpoints with security.
+
+- Secret handling: As part of the API design document we will review a proposed approach for handling secrets securely. The goal of this proposal is to rely on the `secretStores` type and avoid handling secrets as part of other APIs. This will simplify the security model overall for Radius.
+
+- Strong typing for APIs and storage: As described in the [Kubernetes Structural Schema](https://kubernetes.io/blog/2019/06/20/crd-structural-schema/#towards-complete-knowledge-of-the-data-structure) blog post, an API can introduce vulnerabilities whenever it persists unvalidated data. The implementation of user-defined types will increase the level of validation and improve the overall security posture of Radius.
+
+## Compatibility (optional)
+
+Current UCP does not know the individual resource types that belong to a resource provider. It only validates the provider namespace. When we introduce the APIs for registration of resource types/providers, we should begin registering the built-in types. This can be rolled our gradually without breaking users.
+
+## Monitoring and Logging
+
+The Dynamic RP design document will cover monitoring and logging. We don't anticipate any new kinds of telemetry, but if we build a new component we must add the boilerplate to produce the telemetry we already support.
+
+## Development plan
```

Comment:
Thanks for the detailed phasing.. helps understand how this all gets done and rolled out.

Reply:
👍